make ADIOI_GEN_WriteStrided not step on itself

markalle opened the pull request:
The ADIOI_GEN_WriteStrided funcion uses data sieving on non-contiguous
types. That is, if it wants to write data at locations
```    [W...X...Y...Z...]```
it reads the whole buffer
```    [dddddddddddddddd]```
changes the locations it wants to write to
```    [WdddXdddYdddZddd]```
then writes the whole thing back.  It uses locks to make this safe, but
the problem is this only protects against other parts of the product that
are using locks.  And without this PR a peer who is simultaneously making
a simple non-contiguous write wouldn't have locked.

In the fully general case I'm not positive this commit makes WriteStrided
safe as I'm not confident that all WriteContig calls are protected by locks.
But in practice I couldn't come up with a testcase that failed after putting
in this fix.  It turned out that a wide variety of calls end up back at
the location where I added the lock, so a lot of code is made safe by this
checkin.

A testcase to demonstrate the original problem is here:
    https://gist.github.com/markalle/d7da240c19e57f095c5d1b13240dae24
```
% mpicc -o x romio_write_timing.c
% mpirun -np 4 ./x
```
Note: you need to use a filesystem that uses ADIOI_GEN_WriteStrided to
hit the bug.  I was using GPFS.

This commit adds detection at File_set_view() time to see if any of
the ranks are using a non-contiguous view, and uses that to key off of
later in GEN_WriteStrided() to make a rank lock in that case even if
its personal view is contiguous.

Part of this commit is in ad_io_coll.c where a temporary view is created.
I don't have a testcase that goes through that path.  I wasn't sure just
from reading the code whether all calls into ADIOI_IOFiletype() were
collectives, so I didn't want to do an Allreduce there.  So instead I
went a level up and attached an attribute to a temporary datatype that
was created by its caller.  That may have been unnecessary, but that's
how I made sure I was only every running my Allreduce from a location that
I was sure was indeed collective.

Signed-off-by: Mark Allen <markalle@us.ibm.com>

markalle commented:
I'm real curious of romio people's thoughts about both
1. is this PR really a fix, eg does it add enough locking to make sieving safe
2. is the performance acceptable since it should only be increasing the locking when the view is non-contig (at some rank)

markalle commented:
bot:retest

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/warnings
test:mpich/authorship

markalle commented:
My first version of this PR had a number of CI tests (like bigtype.c) fail with the message
```
## File locking failed in ADIOI_GEN_SetLock64(fd C,cmd F_SETLKW/D,type F_WRLCK/3,whence 0) with return value FFFFFFFF and errno 16.
## If the file system is NFS, you need to use NFS version 3, ensure that the lockd daemon is running on all the machines, and mount the directory with the 'noac' option (no attribute caching).
```

And I think to some extent that's a real issue: this PR is potentially adding locking in places that didn't used to have it.  So if a test used to be running on NFS and worked because it didn't used to try acquiring a lock, that test might legitimately fail now.  In the case of the specific tests I saw from the first round of CI though, I noticed they were -np 1 and there's no need to activate my locking code for that case.  So I'm re-pushing with that situation carved out.

hzhou commented:
> So if a test used to be running on NFS and worked because it didn't used to try acquiring a lock, that test might legitimately fail now.

As far as I am aware, our jenkins tests always run on local partitions. I am pinging @roblatham00 for more insights.

hzhou commented:
test:mpich/warnings
test:mpich/ch3/most
test:mpich/ch4/most

roblatham00 commented on src/mpi/romio/adio/common/ad_io_coll.c:
```diff
@@ -223,6 +225,28 @@ void ADIOI_IOStridedColl(ADIO_File fd, void *buf, int count, int rdwr,
         ADIOI_Build_agg_reqs(fd, rdwr, nprocs,
```
See the beginning of ADIOI_GEN_WriteStridedColl line 80 -- are you using system hints or environment variables to set the persistent file realms option?

markalle replied:
There's a possibility I don't actually need the changes in ad_io_coll.c.  I don't believe I had a testcase or runtime situation that used that path, I was just inspecting the code for places where the fd->filetype was set, and made all those places establish a matching fd->view_has_noncontig setting.

markalle replied:
Well, do you think we should keep this change, or could we suppose that operations that occur within one of those persistent file realms where it's resetting fd->filetype are already coordinated and thus wouldn't need any new locking?  If we didn't use the current code I'm not sure if I'd leave fd->view_has_noncontig untouched, or if I'd set it to 0 and restore it where the fd->filetype is being set/restored. Probably the latter.

roblatham00 replied:
I should delete all that file realm code.  Keep it here.  Thanks for explaining the fd->filetype connection.

gpaulsen commented:
It looks like this is one of three ROMIO PRs needed to fix an HDF5 issue on Lustre.  Please see https://github.com/open-mpi/ompi/issues/6871 for details.

roblatham00 commented on src/mpi/romio/adio/common/ad_open.c:
```diff
@@ -62,6 +62,13 @@ MPI_File ADIO_Open(MPI_Comm orig_comm,
+     * it here to fill in the below if it's not necessary. The main place
+     * the below gets checked and set via allreduce is in the set_view call.
+     */
+    fd->view_has_noncontig = 0;
```
correct.  the default file view is indeed contiguous.  MPI_File_set_view will (possibly) alter that

roblatham00 commented on src/mpi/romio/adio/common/ad_write_str.c:
```diff
@@ -280,8 +280,30 @@ void ADIOI_GEN_WriteStrided(ADIO_File fd, const void *buf, int count,
+             * the location they take a lock depending on whether they have
+             * atomicity set, but they end up with a lock either way.
+             */
+            int tsize;
```
The HDF5 test suite triggered a related problem, but what I was seeing was separate independent operations racing ahead of collective operations and getting in the way of the read-modify-write.  My solution was to add a collective pause at the end which can definitely hurt performance.  Look for the Allreduce around line 1033. 

Now obviously that synchronization wasn't enough to fix your test case.  But between your additional locking and my additional synchronization, there is probably a more sophisticated solution we could devise that fixes both.  Not suggesting we spend the next year working on that!  Let's get a fix in place now.

markalle replied:
Sounds good. I remember back when I first put in my extra lock I tried to make related testcases hoping to get other I/O to sneak in while sieving was in place, but I wasn't able to get a failure.  Maybe because of the collective operation you mentioned, or maybe because a lot of write paths end up back where I added the new lock/unlock above.

roblatham00 commented:
It's a good test case.  I don't think we have any other tests that deliberately set up imbalanced operations like that.  I wonder what other areas of ROMIO are buggy in this respect.

markalle commented:
Repushed for a CI failure about compiler warning messages, I added another typecast that I assume will fix it

hzhou commented:
test:mpich/warnings
test:mpich/ch3/most
test:mpich/ch4/most

roblatham00 commented:
@wkliao could you take a look at the change here?   Mark Allen has found a situation where a contiguous write can interfere with a data-sieving write on another process.   Check out that test case demonstrating the problem.   He fixed it by adding locking around the contiguous write if other processes are going to be doing data sieving. 

It has been a while but it was you who identified the performance optimization here --  locking around this write will freeze out everyone else, but they are doing non-contiguous operations anyway so should not expect wire-speed performance.    It could be a very large I/O operation.

wkliao commented:
I believe wrapping a byte-range lock to a contiguous write is sufficient,
when data sieving is enabled. Adding an MPI_Allreduce to check whether
any process's file view is noncontiguous overdoes it, because the successive
writes may not be noncontiguous even if the view is.

roblatham00 commented:
Data sieving writes are only supported on file systems that also support fcntl locks so this approach will work.   On file systems that do not support fcntl locks like PVFS or others, we won't have data sieving to worry about anyway.

wkliao commented:
When data sieving is enabled, an MPI process will do read-modify-write
if its write request is noncontiguous. In order to prevent a different process
from sneaking a contiguous write to the gap of read-modify-write, wrapping
a byte-range-lock to the contiguous write is necessary.

I simply think adding an MPI_Allreduce to check noncontiguous file views
of all process is an overkill.

wkliao commented:
FYI. I was able to run the test program and got the expected error
on Cori at NERSC using Lustre. However, the fix of wrapping byte-range
locks should also be added to `ADIOI_LUSTRE_WriteStrided()` of the
ADIO Lustre driver and maybe other drivers as well.  Once I did that
(just wrapping write with locks), the test program passed.

I also notice that the use of ADIOI_WRITE_LOCK and ADIOI_UNLOCK
when atomic mode is enabled is incorrect, but that will be a separate
issue from this PR.

markalle commented:
@wkliao are you saying the contiguous write lock should be unconditional or have a more targeted condition for when it's activated?

I'd like it to be more targeted about when to activate the contiguous lock, but I don't see what other info the contiguous writer has access to to know that a peer is sieving.  I figured if the view is contiguous at all peers that at least guarantees they're not sieving, so I Allreduced to identify that so the lock could be turned off in that case.  I assume the performance impact would be large if lock was unconditional.  And I don't see what other information could be used to make it more targeted.

As far as which ADIO_WriteContig() calls to protect, that's a big concern I have.  Logically I'd expect the answer is "all of them".  So in my testing I was surprised when the one I added seemed to cover a lot of cases.  Since the lustre and nfs paths both looked really similar to GEN_WriteStrided I added the lock there like you said and repushed.

wkliao commented:
No. I think wrapping the contiguous writes with file locks is
necessary only when data sieving write hint is not disabled,
and this fix is only required by the contiguous writes called
inside MPI independent write functions, not collective.

See my changes in https://github.com/wkliao/mpich/commit/fb42631a5963827e1356524328ab246dbd271a73

markalle commented:
@wkliao Thanks, I like your fix and will plan to close this PR.  But when I was trying to merge your fix I noticed the 3rd change on the ad_write_str.c file is
```
            if (!(fd->atomicity)) ADIOI_UNLOCK(fd, writebuf_off, SEEK_SET, writebuf_len); \
            if (!fd->atomicity && fd->hints->ds_write == ADIOI_HINT_DISABLE) \
                ADIOI_WRITE_LOCK(fd, writebuf_off, SEEK_SET, writebuf_len); \
```
That seems suspicious to be changing from unlock to lock.

hzhou commented:
> No. I think wrapping the contiguous writes with file locks is
> necessary only when data sieving write hint is not disabled,
> and this fix is only required by the contiguous writes called
> inside MPI independent write functions, not collective.
> 
> See my changes in [wkliao@fb42631](https://github.com/wkliao/mpich/commit/fb42631a5963827e1356524328ab246dbd271a73)

@wkliao Have you opened a PR for the mentioned fix?

markalle commented:
I just now did, see if
https://github.com/wkliao/mpich/commit/17354dbd64602bbdb6bfe7cf42ffdc5c3c346ede
looks right

markalle commented:
I got confused on making it a PR with your change not checked in though. Is this PR okay?
https://github.com/pmodels/mpich/pull/4995

markalle commented:
Note: planning to close this PR in favor of https://github.com/pmodels/mpich/pull/4995

wkliao commented:
> @wkliao Thanks, I like your fix and will plan to close this PR. But when I was trying to merge your fix I noticed the 3rd change on the ad_write_str.c file is
> 
> ```
>             if (!(fd->atomicity)) ADIOI_UNLOCK(fd, writebuf_off, SEEK_SET, writebuf_len); \
>             if (!fd->atomicity && fd->hints->ds_write == ADIOI_HINT_DISABLE) \
>                 ADIOI_WRITE_LOCK(fd, writebuf_off, SEEK_SET, writebuf_len); \
> ```
> 
> That seems suspicious to be changing from unlock to lock.

Thanks for catching my copy-and-paste mistake.

markalle commented:
@wkliao can you review https://github.com/pmodels/mpich/pull/4995 then and I guess after that I'll close this one.

wkliao commented:
It looks fine with me, thanks.
